### DISCUSSION TITLE
resuming specific workflow instances

### DISCUSSION THREAD

**Discussion by ziedunited92** [elsa-workflows/elsa-core#637] [Q&A] [answered]
I am building an Elsa workflow for reports , my reports have a uniform workflow . my problem is that when resuming a special workflow of a special report i want it to get the id of the workflow as a parameter to resume exactly the target workflow not all of them.

**Comment by AdelSS04** [comment]
i have the same question ...

**Comment by ziedunited92** [comment]
can you help me please @sfmskywalker

**Comment by sfmskywalker** [comment, marked as answer]
Of course. In Elsa 1, you can resume specific workflow instances with `IWorkflowInvoker.ResumeAsync`. See [Sample 25](https://github.com/elsa-workflows/elsa-core/blob/master/src/samples/Sample25/Program.cs#L47).

**Reply by ziedunited92** [reply]
how can i make it work Via designer ? ,and does it work with the current version ?

**Reply by sfmskywalker** [reply]
I'm not sure I understand what you mean by making it work via the designer. What specifically would you like to do from the designer?

The designer is used to design workflows. Those workflows get executed & resumed by the workflow host for you. 

For example, if you add a Timer activity as a starting node, the workflow will execute at the specified interval. If you add a Timer activity as a node somewhere later in the workflow, the workflow instance will initially be suspended when it first hits that activity, and when the timer reaches the specified interval, the workflow gets resumed automatically.

If you are designing your own trigger activity, it is up to you to start/resume workflows that start with or are blocked on that activity. You can find various examples of this in the form of the Timer and HTTP Request activities in Elsa 1, and Azure Service Bus and Rebus activities in Elsa 2. This does not involve the designer itself; the user would only add your custom trigger, and the auxiliary code around it will take care of the rest.

**Reply by ziedunited92** [reply]
Well, i have this workflow for example 
![image](https://user-images.githubusercontent.com/50137270/108054582-21feb500-704f-11eb-9a6c-2daa921c1993.png)

when i reach the first end point : Recive Http Request(/adel) i got an executing instance. if do it again, i got 2 Executing instances.
here evrything is working well , but if i reach the final end point (zied) , i got 2 finished instance.
what i want is having the instance id as a parameter to finish only the target workflow not both  .
![image](https://user-images.githubusercontent.com/50137270/108055373-2e374200-7050-11eb-8849-131d4a4b1803.png)

**Reply by sfmskywalker** [reply]
Aha, I understand now. What you need is a way to make requests that can target a specific workflow instance.

What you need to do is this:

1. Assign a correlation ID to the workflow when it executes.
2. When making the second HTTP request (to /zied), supply the correlation ID that was assigned.

In Elsa 1, you would do the above as follows:

1. Add the Correlate activity next to the first activity (connect Receive HTTP Request to Correlate, and connect Correlate to the Write HTTP Response activity).
2. Use the following JavaScript expression to generate a correlation ID when configuring the Correlate activity: `(newGuid())`. Another approach is to simply specify some correlation ID via some querystring parameter when making the initial request, e.g. ?myId=foo, and use the following JS expression to access that value when assigning the correlation ID: `(queryString('myId'))`.
3. In case you went for generating a correlation ID using `newGuid()`, you can make it easy on yourself to get the generated correlation ID by updating **Write HTTP Response** to write it out using the following JavaScript syntax: `(correlationId())`. This should write out the correlation ID to the response body. I thought there was a liquid filter for this, but I couldn't find it - this is something I'll add to Elsa 2, since liquid is easier to work with when constructing HTTP response bodies compared to concating strings in JS.
4. Perform an HTTP GET request to /zied with the following request header: `X-Correlation-Id`, specifying the appropriate correlation ID.

In Elsa 2, you can specify the correlation ID via both header as well as query string parameter.

**Reply by sfmskywalker** [reply]
Note to self: consider allowing the workflow instance ID itself to be used as a "correlation id" when sending an HTTP request. Would simplify the sample workflow.

**Reply by ziedunited92** [reply]
First of all i have to thank you very much on taking the time to answer my questions and i am sorry i didn't resolve my problem yet .
to clarify i'am using Elsa 1 .
I have added the correlate activity as you have mentionned
![image](https://user-images.githubusercontent.com/50137270/108060555-4fe7f780-7057-11eb-97d5-3145719d4d67.png)
![image](https://user-images.githubusercontent.com/50137270/108060617-642bf480-7057-11eb-9f7c-a4b58d7126b7.png)
![image](https://user-images.githubusercontent.com/50137270/108060708-8aea2b00-7057-11eb-84ef-1c40e847b409.png)
![image](https://user-images.githubusercontent.com/50137270/108060829-b836d900-7057-11eb-956e-e99fcba968ec.png)
and both workflows are finished
![image](https://user-images.githubusercontent.com/50137270/108060933-e3212d00-7057-11eb-9f3f-316778dda3a4.png)

**Reply by ziedunited92** [reply]
I'am blocked can you answer me please

**Reply by sfmskywalker** [reply]
Could you send me your workflow definition file (exported as JSON) so I can take a look?

**Reply by ziedunited92** [reply]
`{"activities":[{"id":"c0aba00a-6b65-44b3-a4a8-51c57c2ba422","type":"Correlate","left":214,"top":273,"state":{"valueExpression":{"expression":"(newGuid())","syntax":"JavaScript"},"name":"","title":"","description":""},"blocking":false,"executed":false,"faulted":false},{"id":"113400b5-4010-4885-9312-275e30c798c1","type":"ReceiveHttpRequest","left":1038,"top":86,"state":{"path":"/zied2","method":"GET","readContent":"true","name":"","title":"","description":""},"blocking":false,"executed":false,"faulted":false},{"id":"f72e48b6-1a77-48c9-a0ff-3cceaf3fc56a","type":"ReceiveHttpRequest","left":36,"top":34,"state":{"path":"/adel","method":"GET","readContent":null,"name":"","title":"","description":""},"blocking":false,"executed":false,"faulted":false},{"id":"29a85c8e-2ff9-4504-b8eb-86925accf017","type":"TriggerSignal","left":1534,"top":574,"state":{},"blocking":false,"executed":false,"faulted":false},{"id":"3b990e8e-fe5d-4412-b8f1-6a8b0a2d972c","type":"TriggerSignal","left":1636,"top":511,"state":{},"blocking":false,"executed":false,"faulted":false},{"id":"98310b55-ce8f-47f5-8ebf-8cea12302efb","type":"WriteHttpResponse","left":434,"top":128,"state":{"statusCode":"200","content":{"expression":"hello!","syntax":"Literal"},"contentType":"application/json","responseHeaders":{"expression":"x-powered-by:Elsa Workflows","syntax":"Literal"},"name":"","title":"","description":""},"blocking":false,"executed":false,"faulted":false}],"connections":[{"sourceActivityId":"c0aba00a-6b65-44b3-a4a8-51c57c2ba422","destinationActivityId":"98310b55-ce8f-47f5-8ebf-8cea12302efb","outcome":"Done"},{"sourceActivityId":"98310b55-ce8f-47f5-8ebf-8cea12302efb","destinationActivityId":"113400b5-4010-4885-9312-275e30c798c1","outcome":"Done"},{"sourceActivityId":"f72e48b6-1a77-48c9-a0ff-3cceaf3fc56a","destinationActivityId":"c0aba00a-6b65-44b3-a4a8-51c57c2ba422","outcome":"Done"}]}`

here it is

**Reply by ziedunited92** [reply]
can you answer me please it's the 4th day i'am blocked in the same point.

**Reply by sfmskywalker** [reply]
Thank you for providing the JSON, it's helpful. Sorry you have been stuck for so long. I wish I had more time to assist, but it's not always easy to find the time. But, rest assured that I'm looking into this today.

**Reply by ziedunited92** [reply]
thank you

**Reply by sfmskywalker** [reply]
I just tried, but it works fine on my end:

1. I imported your JSON and published the workflow.
2. I made multiple requests to /adel in order to generate multiple workflow instances
3. I took the correlation ID of one of the workflow instances and made a new request to /zied2 while providing the X-Correlation-Id header.

The result was exactly as expected: the workflow instance with the specified correlation ID was finished, while the others remained in the status Executing.

As a next step to try and troubleshoot this, I suggest you either:

1) Stop the application, delete all workflow instances, and try again.
2) If you get the same results, send me your application + database, and hopefully then I can reproduce the issue you are seeing.

**Reply by sfmskywalker** [reply]
An additional thing you might try is to clone the source code and launch the Elsa.Dashboard.Web project from there, and see if you can reproduce the issue. I'm assuming you are referencing the latest Elsa 1.5 packages in your project, but if not, it is possible that you are referencing older versions that don't yet have the correlation handling in place.

**Reply by ziedunited92** [reply]
Thank you very much that worked now seems that's a version issue .
Just I want to ask you one more question please .
How can i make the correlation Id necessary to finish the instance because now if i don't inject a correlation id all instance are finished.

**Reply by sfmskywalker** [reply]
Unfortunately there's no way to do this in Elsa 1 without refactoring the HTTP activities middleware. In Elsa 2 this has been fixed, but if you need it fixed in Elsa 1 I am open to receiving a PR.

**Reply by ziedunited92** [reply]
If i have a proposition i will make you know.
But i have a question : can i create a custom activity that works like the receive http request with the path and method proprities are predefined

**Reply by sfmskywalker** [reply]
Most certainly. Could be as simple as creating your own activity and triggering workflows from some API controller. It would work similar to the existing ReceiveHttpRequest activity. The primary difference might be that it relies on middleware to capture requests and trigger workflows if you choose to use a controller.

**Reply by ziedunited92** [reply]
I'am sorry , I didn't understand. you mean i use the  "trigger workflow" activity ?

**Reply by sfmskywalker** [reply]
No, you create your own custom activity, and use `IWorkflowInvoker` to trigger workflows that start with or are blocked on your custom activity.

**Reply by Nhollas** [reply]
Hi @sfmskywalker, is there an example to the above? It appears the previous link you sent is no longer available.. :(

**Reply by sfmskywalker** [reply]
Sure, here you go: https://github.com/elsa-workflows/elsa-core/blob/1.4.x/src/samples/Sample25/Program.cs#L47
